hls4ml/model/optimizer/passes/infer_precision.py

```diff
@@ -0,0 +1,290 @@
+            node.weights['pointwise'].update_precision(node.types['pointwise_t'].precision)
+
+            inferred_types.append('pointwise_t')
+        else:
+            pointwise_width = node.types['pointwise_t'].precision.width
+            pointwise_integers = node.types['pointwise_t'].precision.integer
+
+        if 'bias_t' in types_to_infer:
+            bias_quantizer = node.get_attr('bias_quantizer', None)
+            if bias_quantizer is not None:
+                bias_width = bias_quantizer.bits
+                bias_integers = bias_quantizer.hls_type.integer
+                node.types['bias_t'].name = node.name + '_bias_t'
+                node.types['bias_t'].precision = bias_quantizer.hls_type
+            else:
+                self._infer_default_type(node, 'bias_t')
+                bias_width = node.types['bias_t'].precision.width
+                bias_integers = node.types['bias_t'].precision.integer
+            node.weights['bias'].update_precision(node.types['bias_t'].precision)
+
+            inferred_types.append('bias_t')
+        else:
+            bias_width = node.types['bias_t'].precision.width
+            bias_integers = node.types['bias_t'].precision.integer
+
+        n_ops = node.get_attr('n_chan')
+        new_type = FixedPrecisionType(
+            width=int(max(np.ceil(input_width + pointwise_width + np.log2(n_ops)), bias_width) + 1),
+            integer=int(max(np.ceil(input_integers + pointwise_integers + np.log2(n_ops)), bias_integers) + 1),
+        )
+
+        if 'accum_t' in types_to_infer:
+            node.types['accum_t'].name = node.name + '_accum_t'
+            node.types['accum_t'].precision = new_type
+
+            inferred_types.append('accum_t')
+
+        if 'result_t' in types_to_infer:
+            node.types['result_t'].name = node.name + '_result_t'
+            node.types['result_t'].precision = new_type
+
+            inferred_types.append('result_t')
+
+        return inferred_types
+
+    def _infer_bn_precision(self, node, types_to_infer):
+        inferred_types = []
+
+        if 'scale_t' in types_to_infer:
+            self._infer_default_type(node, 'scale_t')
+            node.weights['scale'].update_precision(node.types['scale_t'].precision)
+            inferred_types.append('scale_t')
+
+        if 'bias_t' in types_to_infer:
+            self._infer_default_type(node, 'bias_t')
+            node.weights['bias'].update_precision(node.types['bias_t'].precision)
+            inferred_types.append('bias_t')
+
+        if 'result_t' in types_to_infer:
+            scale_precision = node.types['scale_t'].precision
+            bias_precision = node.types['bias_t'].precision
+
+            out_precision = deepcopy(node.get_input_node().get_output_variable().type.precision)
+            out_precision.integer += scale_precision.integer
+            out_precision.fractional = max(out_precision.fractional, scale_precision.fractional)
```

Comment:
Explicitly keeping integer, fractional, and consistent worries me. I would propose making fractional a property of the `FixedPrecisionType` like:
```python
    @property
    def fractional(self):
        self.width - self.integer
```

Also, another concern with this logic is if the input node type is an `IntegerPrecionsType`. Setting fractional and integer explicitly could result in a misformed type. I would propose explicitly creating a new `FixedPrecisionType` vs editing the existing one.